File: crates/benches/fixtures/dependency_graph.py
from __future__ import annotations

import ast
import sys
from collections import defaultdict, deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterator, Optional


@dataclass
class Module:
    name: str
    path: Path
    imports: list[str] = field(default_factory=list)


class DependencyGraph:
    def __init__(self) -> None:
        self._nodes: dict[str, Module] = {}
        self._edges: dict[str, set[str]] = defaultdict(set)
        self._reverse: dict[str, set[str]] = defaultdict(set)

    def add_module(self, module: Module) -> None:
        self._nodes[module.name] = module
        for imp in module.imports:
            self._edges[module.name].add(imp)
            self._reverse[imp].add(module.name)

    def dependencies(self, name: str) -> set[str]:
        return self._edges.get(name, set())

    def dependents(self, name: str) -> set[str]:
        return self._reverse.get(name, set())

    def transitive_deps(self, name: str) -> set[str]:
        visited: set[str] = set()
        queue = deque([name])
        while queue:
            current = queue.popleft()
            for dep in self._edges.get(current, set()):
                if dep not in visited:
                    visited.add(dep)
                    queue.append(dep)
        return visited

    def impact_set(self, name: str) -> set[str]:
        visited: set[str] = set()
        queue = deque([name])
        while queue:
            current = queue.popleft()
            for dep in self._reverse.get(current, set()):
                if dep not in visited:
                    visited.add(dep)
                    queue.append(dep)
        return visited

    def cycles(self) -> list[list[str]]:
        visited: set[str] = set()
        rec_stack: set[str] = set()
        found: list[list[str]] = []

        def dfs(node: str, path: list[str]) -> None:
            visited.add(node)
            rec_stack.add(node)
            for neighbor in self._edges.get(node, set()):
                if neighbor not in visited:
                    dfs(neighbor, path + [neighbor])
                elif neighbor in rec_stack:
                    found.append(path + [neighbor])
            rec_stack.discard(node)

        for node in list(self._nodes):
            if node not in visited:
                dfs(node, [node])
        return found

File: crates/benches/fixtures/dependency_graph.go
package depgraph

import (
	"fmt"
	"sort"
)

// Module represents a code module with its dependencies.
type Module struct {
	Name    string
	Path    string
	Imports []string
}

// DependencyGraph holds the import relationship graph.
type DependencyGraph struct {
	nodes   map[string]*Module
	edges   map[string][]string
	reverse map[string][]string
}

// NewDependencyGraph initializes an empty graph.
func NewDependencyGraph() *DependencyGraph {
	return &DependencyGraph{
		nodes:   make(map[string]*Module),
		edges:   make(map[string][]string),
		reverse: make(map[string][]string),
	}
}

// Add registers a module and its imports.
func (g *DependencyGraph) Add(m *Module) {
	g.nodes[m.Name] = m
	for _, imp := range m.Imports {
		g.edges[m.Name] = append(g.edges[m.Name], imp)
		g.reverse[imp] = append(g.reverse[imp], m.Name)
	}
}

// Dependents returns direct dependents of the given module.
func (g *DependencyGraph) Dependents(name string) []string {
	return g.reverse[name]
}

// ImpactSet computes the transitive set of modules affected by a change.
func (g *DependencyGraph) ImpactSet(name string) []string {
	visited := make(map[string]bool)
	queue := []string{name}
	for len(queue) > 0 {
		cur := queue[0]; queue = queue[1:]
		for _, dep := range g.reverse[cur] {
			if !visited[dep] {
				visited[dep] = true
				queue = append(queue, dep)
			}
		}
	}
	result := make([]string, 0, len(visited))
	for k := range visited { result = append(result, k) }
	sort.Strings(result)
	return result
}

// TopologicalOrder returns a topological sort, or an error if a cycle exists.
func (g *DependencyGraph) TopologicalOrder() ([]string, error) {
	inDegree := make(map[string]int)
	for name := range g.nodes { inDegree[name] = 0 }
	for _, targets := range g.edges {
		for _, t := range targets { inDegree[t]++ }
	}
	var queue []string
	for name, deg := range inDegree { if deg == 0 { queue = append(queue, name) } }
	sort.Strings(queue)
	var order []string
	for len(queue) > 0 {
		cur := queue[0]; queue = queue[1:]
		order = append(order, cur)
		for _, dep := range g.edges[cur] {
			inDegree[dep]--
			if inDegree[dep] == 0 { queue = append(queue, dep); sort.Strings(queue) }
		}
	}
	if len(order) != len(g.nodes) { return nil, fmt.Errorf("cycle detected") }
	return order, nil
}
